spackbot/pr_add_reviewers.py

```diff
@@ -0,0 +1,235 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import contextlib
+import logging
+import os
+import re
+import tempfile
+
+import gidgethub
+from sh.contrib import git
+from gidgethub import routing
+
+logger = logging.getLogger(__name__)
+router = routing.Router()
+
+package_path = r"^var/spack/repos/builtin/packages/(\w[\w-]*)/package.py$"
+
+non_reviewers_comment = """\
+  @{non_reviewers} can you review this PR?
+
+  This PR modifies the following package(s), for which you are listed as a maintainer:
+
+  * {packages_with_maintainers}
+"""
+
+no_maintainers_comment = """\
+Hi @{author}! I noticed that the following package(s) don't yet have maintainers:
+
+* {packages_without_maintainers}
+
+Are you interested in adopting any of these package(s)? If so, simply add the following to the package class:
+```python
+    maintainers = ['{author}']
+```
+If not, could you contact the developers of this package and see if they are interested? Please don't add maintainers without their consent.
+
+_You don't have to be a Spack expert or package developer in order to be a "maintainer", it just gives us a list of users willing to review PRs or debug issues relating to this package. A package can have multiple maintainers; just add a list of GitHub handles of anyone who wants to volunteer._
+"""
+
+
+async def changed_packages(gh, pull_request):
+    """Return an array of packages that were modified by a PR.
+
+    Ignore deleted packages, since we can no longer query them for
+    maintainers.
+
+    """
+    # see which files were modified
+    packages = []
+    async for f in gh.getiter(pull_request["url"] + "/files"):
+        filename = f["filename"]
+        status = f["status"]
+
+        if status == "removed":
+            #            print(f["patch"])
+            continue
+
+        match = re.match(package_path, filename)
+        if not match:
+            continue
+        packages.append(match.group(1))
+
+    return packages
+
+
+@contextlib.contextmanager
+def temp_dir():
+    """Create a temporary directory, cd into it, destroy it and cd back when done."""
+    pwd = os.getcwd()
+    with tempfile.TemporaryDirectory() as temp_dir:
+        try:
+            os.chdir(temp_dir)
+            yield temp_dir
+        finally:
+            os.chdir(pwd)
+
+
+def find_maintainers(packages, repository, pull_request, number):
+    """Return an arrays of packages with maintainers, an array of packages
+    without maintainers, and a set of maintainers.
+
+    Ignore the author of the PR, as they don't need to review their own PR.
```

Comment:
This could use a better docstring with type annotations and things. Actually, now that we don't have to support old Python, how do you feel about mypy? I've been using it for all of my projects. It hasn't yet caught much when my code is already covered by unit tests, but since we don't have unit tests (yet) it may actually help prevent bugs.

Reply:
can we hold off on the mypy and maybe add that later?

Reply:
I have mixed feelings about it.  I've seen a bunch of people struggle to understand how to do the type annotations, though that may be because in Spack they have to be in comments.

Reply:
We can def hold off on that